src/charm.py

```diff
@@ -2,134 +2,108 @@
-        name="envoy.http_connection_manager",
-        config=hcm.HttpConnectionManager(
-            codec_type=hcm.HttpConnectionManagerCodecType.AUTO,
-            stat_prefix="ingress_http",
-            route_config=api.RouteConfiguration(
-                name="local_route",
-                virtual_hosts=[virtual_host],
+        # TODO: Change this from SDI to the service-info lib
+        self.grpc = self.charm_reconciler.add(
+            component=SdiRelationDataReceiverComponent(
+                charm=self,
+                name="relation:grpc",
+                relation_name="grpc",
             ),
-            http_filters=[
-                hcm.HttpFilter(name="envoy.grpc_web"),
-                hcm.HttpFilter(name="envoy.cors"),
-                hcm.HttpFilter(name="envoy.router"),
-            ],
-        ),
-    )
-    return api.Listener(
-        name="listener_0",
-        address=api.core.Address(
-            socket_address=api.core.SocketAddress(
-                address="0.0.0.0",
-                port_value=port,
-            )
-        ),
-        filter_chains=[api.listener.FilterChain(filters=[filter_listener])],
-    )
-
-
-class CheckFailed(Exception):
-    """Raise this exception if one of the checks in main fails."""
+            depends_on=[self.leadership_gate],
+        )
 
-    def __init__(self, msg: str, status_type=None):
-        super().__init__()
+        # Should this Component block the charm if it is not available?
+        # It is a requirement of the Kubeflow bundle, not the envoy charm.
+        # Envoy doesn't  NEED this relation, but Charmed Kubeflow using this for
+        # kfp does need that relation.
+        self.ingress_relation = self.charm_reconciler.add(
+            component=SdiRelationBroadcasterComponent(
+                charm=self,
+                name="relation:ingress",
+                relation_name="ingress",
+                data_to_send={
+                    "prefix": "/ml_metadata",
+                    "rewrite": "/ml_metadata",
+                    "service": self.model.app.name,
+                    "port": int(self.model.config["http-port"]),
+                },
+            ),
+            depends_on=[self.leadership_gate],
+        )
 
-        self.msg = str(msg)
-        self.status_type = status_type
-        self.status = status_type(self.msg)
+        self.envoy_config_generator = self.charm_reconciler.add(
+            GenerateEnvoyConfig(
+                charm=self,
+                name="envoy_config_generator",
+                inputs_getter=lambda: GenerateEnvoyConfigInputs(
+                    admin_port=int(self.config["admin-port"]),
+                    http_port=int(self.config["http-port"]),
+                    # TODO: Do these work properly?  They're not deferred
+                    upstream_service=self.grpc.component.get_data()["service"],
+                    upstream_port=self.grpc.component.get_data()["port"],
```

Comment:
You're right that `get_data()` will return empty inputs if its executed without any data on the other end.  I don't think this can happen in the charm because `self.envoy_config_generator` `depends_on=[self.grpc]`, so the grpc component must go Active before `envoy_config_generator` ever attempts to use `get_data()`.  

But, just in case it can, if `get_data()` was called and returned empty inputs, you're right that `GenerateEnvoyConfig` would break here.  In the version you reviewed, `GenerateEnvoyConfig` (and thus the charm) would go into Blocked, but not with such a helpful error message.  To improve this, I've:
* added `test_envoy_config_generator_if_grpc_get_data_returns_empty_dict` which asserts the config generator and charm are not active if malformed inputs are passed to the config generator
* changed error messaging around handling the errors so the logs explain the failure more

wdyt?